Allowing the Run class to access wtg_parents

SHZ66 opened the pull request:
**Issue Number.** Is this pull request related to any outstanding issues? If so, list the issue number.  
Currently, the `Run` class does not have a direct interface to retrieve `wtg_parents` from `west.h5`.  

**Describe the changes made.** A clear and concise description of what the problem is and what you did to fix it. E.g. [...] was happening and I've changed [...] to fix it.  
This PR implements interfaces in `Run` which allows it to directly access `wtg_parents`. It also implements a `wtg_children` function which is an equivalent of the `children` property for retrieving all children according to the graph. 

**Major files changed.**  
- [x] src/westpa/analysis/core.py

**Status.**
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] I have read the CONTRIBUTING document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

**Additional context.** Add any other context or screenshots about the pull request here.

SHZ66 committed:
Merge pull request #25 from westpa/westpa-2.0-restruct

Westpa 2.0 restruct

Commit:
allow load_plugins to take argments that contain plugin configs directly

Commit:
Allow specifying RecursiveBinMapper in west.cfg

Commit:
removed debug message

Commit:
process the copy of the bin setup instead of the original to avoid config being removed

Commit:
making data manager to use its own system reference other than the global one

Commit:
Merge branch 'main' into westpa2

SHZ66 committed:
Merge pull request #50 from westpa/westpa2

update

Commit:
allowing Run to access wtg parents and children

Commit:
critical bug fix to data manager when getting a subset of the segments

SHZ66 committed:
Merge pull request #54 from westpa/develop

update

jeremyleung521 reviewed (approved):
Tested and LGTM. Would be nice to have some CI tests for `westpa.analysis` in general but that's out of the scope of this PR.

jpthompson17 commented on src/westpa/analysis/core.py:
```diff
@@ -530,6 +526,30 @@ def parent(self):
+
+    @property
+    def wtg_parents(self):
+        """Iterable[Walker or InitialState]: The wtg parents of the walker."""
```
or -> |

Commit:
added an option to wtg related functions to enabling/disabling recycling check

SHZ66 commented:
A few changes according to feedback from @jpthompson17:
- Changed WTG related properties to functions (since they can be a bit expensive to run);
- Added an option to WTG related functions to enable/disable the recycling check. 
- Added docs string to spell out WTG and explain what the option does.

Commit:
removed trailing space

Commit:
remove trailing space

jpthompson17 commented on src/westpa/analysis/core.py:
```diff
@@ -542,9 +542,23 @@ def _get_wtg_parent_ids(self):
-        """Iterable[Walker or InitialState]: The wtg parents of the walker."""
+    def get_wtg_parents(self, allow_recycled=True):
+        """
+        Iterable[Walker or InitialState]: The parents of the walker according to 
```
Since this is now a method, we should move the return type `Iterable[Walker | InitialState]` to the "Returns" section of the docstring.

jpthompson17 commented on src/westpa/analysis/core.py:
```diff
@@ -542,9 +542,23 @@ def _get_wtg_parent_ids(self):
+            when the parents were merged and recycled in the last iteration.
+            Set to ``False`` to disallow returning weight transfer parents if 
+            the current walker was recycled in the last iteration.
+        """
```
Add newline before closing """.

jpthompson17 commented on src/westpa/analysis/core.py:
```diff
@@ -542,9 +542,23 @@ def _get_wtg_parent_ids(self):
+        """
+
+        if not allow_recycled and self.initial:
+            return
```
Should this return an empty list/tuple? If not, we may want to update the return type to `Optional[Iterable[Walker | InitialState]]`.

SHZ66 replied:
This function is a generator so in the case where it needs to return empty I can only call `return` without anything.

jpthompson17 commented on src/westpa/analysis/core.py:
```diff
@@ -542,9 +542,23 @@ def _get_wtg_parent_ids(self):
+            A walker that is just created can have weight transfer parents
+            when the parents were merged and recycled in the last iteration.
+            Set to ``False`` to disallow returning weight transfer parents if 
+            the current walker was recycled in the last iteration.
```
I'm wondering if this could be made more explicit, e.g.,
```
Parameters
-----------
include_recycling_events : bool, default True
    Whether to include weight transfers that coincide with recycling events.
```

Commit:
docs and parameter name changes

SHZ66 commented:
@jpthompson17 Thanks for reviewing! Please check out the latest commit. I fixed everything according to the feedback except for one item.

SHZ66 committed:
Merge pull request #56 from westpa/develop

Develop
